quinn-proto/src/endpoint.rs

```diff
@@ -597,15 +597,51 @@ impl Endpoint {
         trace!(self.ctx.log, "connection got packet"; "connection" => %self.connections[conn.0].local_id, "len" => packet.payload.len());
         let was_closed = self.connections[conn.0].state.as_ref().unwrap().is_closed();
 
+        let mut err_to_log: Option<ConnectionError> = None;
         // State transitions
         let state = self.connections[conn.0].state.take().unwrap();
-        let state = self.connections[conn.0].handle_connected_inner(
+        let state = match self.connections[conn.0].handle_connected_inner(
             &mut self.ctx,
             now,
             remote,
             packet,
             state,
-        );
+        ) {
+            Ok(state) => match state {
+                State::Closed(closed) => {
+                    let copy = closed.clone();
+                    err_to_log = Some(closed.reason.into());
+                    State::Closed(copy)
+                }
+                _ => state,
+            },
+            Err(conn_err) => match conn_err {
+                ConnectionError::TransportError { error_code } => {
+                    err_to_log = Some(error_code.into());
+                    State::handshake_failed(error_code, None)
```

Comment:
We should only enter `HandshakeFailed` if we were initially in `Handshake`. If the connection was established, we should enter `Closed`.

Reply:
Got it.  By being in this arm (`ConnectionError::TransportError`), do we know that a connection was established?  If so, I'll just change this to `State::Closed`.  Or do we have to check the previous state for `Handshake` from the initial `let state = self.connections[conn.0].state.take().unwrap();` line?

Reply:
> in the previous implementation, every time we logged a ConnectionError::TransportError, it returned a State::handshake_failed.

This is incorrect; for example, examine `Connection::process_payload`.

> By being in this arm (ConnectionError::TransportError), do we know that a connection was established?

Transport errors can arise at any time, so you'll want to inspect the previous state, yeah.

Reply:
Alright.  If previous state was `State::Handshake`, we will return `State::handshake_failed(...)`.  Otherwise we'll return `State::closed(...)`.